### ISSUE TITLE
[RLlib] Check reparameterization trick for squashed gaussian 

### ISSUE THREAD

**Issue by gresavage** [ray-project/ray#40321]
### What happened + What you expected to happen

For context see: https://discuss.ray.io/t/malformed-reparameterization-trick-in-squashed-gaussian/9651/3

Raised here as an issue at the behest of @ArturNiederfahrenhorst 

It looks like the "reparameterization trick" is implemented slightly differently in RLlib vs. literature.

Literature:
$a = \tanh ({\mu_\theta + \sigma_\theta\odot\xi})$ where $\xi\thicksim N(0,I)$, then mapping to $[low,high]$

RLlib:
$a = \tanh ({a_n})$ where $a_n\thicksim N(\mu_\theta,\sigma_\theta\odot I)$ then mapping to $[low,high]$

It is not clear whether these are mathematically equivalent. From the attached tensorboard summary the gradient norms have slightly different magnitude, and the $\alpha$-loss is different, but otherwise the data is not significantly different.

### Versions / Dependencies

Ray 2.7 (present since at least 2.0)
Python 3.10 (originally discovered with Python 3.8)
Ubuntu 22.04

### Reproduction script

```python
from pprint import pprint

import gymnasium as gym
import numpy as np
import torch
from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
from ray.rllib.algorithms.dqn import DQN
from ray.rllib.algorithms.sac import SAC, SACConfig
from ray.rllib.algorithms.sac.sac_tf_policy import SACTFPolicy, build_sac_model, postprocess_trajectory, validate_spaces
from ray.rllib.algorithms.sac.sac_torch_policy import (
    ComputeTDErrorMixin,
    action_distribution_fn,
    actor_critic_loss,
    optimizer_fn,
    setup_late_mixins,
    stats,
)
from ray.rllib.models.catalog import ModelCatalog
from ray.rllib.models.modelv2 import ModelV2
from ray.rllib.models.torch.torch_action_dist import (
    TorchBeta,
    TorchCategorical,
    TorchDiagGaussian,
    TorchDirichlet,
    TorchDistributionWrapper,
    TorchSquashedGaussian,
)
from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
from ray.rllib.policy.policy import Policy
from ray.rllib.policy.policy_template import build_policy_class
from ray.rllib.policy.torch_mixins import TargetNetworkMixin
from ray.rllib.utils import MAX_LOG_NN_OUTPUT, MIN_LOG_NN_OUTPUT, SMALL_NUMBER, deprecation_warning, override
from ray.rllib.utils.spaces.simplex import Simplex
from ray.rllib.utils.torch_utils import apply_grad_clipping, concat_multi_gpu_td_errors
from ray.rllib.utils.typing import AlgorithmConfigDict, TensorType
from ray.util.debug import log_once


def _get_dist_class(policy: Policy, config: AlgorithmConfigDict, action_space: gym.spaces.Space) -> type[TorchDistributionWrapper]:
    """Helper function to return a dist class based on config and action space.

    Args:
        policy: The policy for which to return the action
            dist class.
        config: The Algorithm's config dict.
        action_space (gym.spaces.Space): The action space used.

    Returns
    -------
        Type[TFActionDistribution]: A TF distribution class.
    """
    if hasattr(policy, "dist_class") and policy.dist_class is not None:
        return policy.dist_class
    elif config["model"].get("custom_action_dist"):
        action_dist_class, _ = ModelCatalog.get_action_dist(action_space, config["model"], framework="torch")
        return action_dist_class
    elif isinstance(action_space, gym.spaces.Discrete):
        return TorchCategorical
    elif isinstance(action_space, Simplex):
        return TorchDirichlet
    else:
        assert isinstance(action_space, gym.spaces.Box)
        if config["normalize_actions"]:
            return TorchModSquashedGaussian if not config["_use_beta_distribution"] else TorchBeta
        else:
            return TorchDiagGaussian


class TorchModSquashedGaussian(TorchSquashedGaussian):
    def __init__(self, inputs: list[TensorType], model: TorchModelV2, low: float = -1.0, high: float = 1.0):
        """Parameterizes the distribution via `inputs`.

        Args:
            low: The lowest possible sampling value
                (excluding this value).
            high: The highest possible sampling value
                (excluding this value).
        """
        if log_once("torch_action_dist_squashed_gaussian_deprecation"):
            deprecation_warning(old="ray.rllib.models.torch.torch_action_dist.TorchSquashedGaussian")
        super().__init__(inputs, model)
        # Split inputs into mean and log(std).
        mean, log_std = torch.chunk(self.inputs, 2, dim=-1)
        # Clip `scale` values (coming from NN) to reasonable values.
        log_std = torch.clamp(log_std, MIN_LOG_NN_OUTPUT, MAX_LOG_NN_OUTPUT)
        std = torch.exp(log_std)
        # reparameterization trich uses N(0,1)
        self.dist = torch.distributions.normal.Normal(torch.zeros_like(mean), torch.ones_like(std))
        assert np.all(np.less(low, high))
        self.low = low
        self.high = high
        self.mean = mean
        self.std = std

    def sample(self) -> TensorType:
        normal_sample = self.dist.rsample()
        normal_sample = self.reparameterize(normal_sample)
        self.last_sample = self._squash(normal_sample)
        return self.last_sample

    def logp(self, x: TensorType) -> TensorType:
        # Unsquash values (from [low,high] to ]-inf,inf[)
        unsquashed_values = self._unsquash(x)
        unparamed_values = self.unparameterize(unsquashed_values)
        # Get log prob of unsquashed values from our Normal.
        log_prob_gaussian = self.dist.log_prob(unparamed_values)
        # For safety reasons, clamp somehow, only then sum up.
        log_prob_gaussian = torch.clamp(log_prob_gaussian, -100, 100)
        log_prob_gaussian = torch.sum(log_prob_gaussian, dim=-1)
        # Get log-prob for squashed Gaussian.
        unsquashed_values_tanhd = torch.tanh(unsquashed_values)
        return log_prob_gaussian - torch.sum(
            torch.log(1 - unsquashed_values_tanhd**2 + SMALL_NUMBER), dim=-1,
        )

    def sample_logp(self):
        z = self.dist.rsample()
        actions = self.reparameterize(z)
        actions = self._squash(actions)
        return actions, torch.sum(self.dist.log_prob(z) - torch.log(1 - actions * actions + SMALL_NUMBER), dim=-1)

    def reparameterize(self, raw_values: TensorType) -> TensorType:
        return self.mean + raw_values * self.std

    def unparameterize(self, values: TensorType) -> TensorType:
        return (values - self.mean) / (self.std + SMALL_NUMBER)


def build_mod_sac_model_and_action_dist(
    policy: Policy,
    obs_space: gym.spaces.Space,
    action_space: gym.spaces.Space,
    config: AlgorithmConfigDict,
) -> tuple[ModelV2, type[TorchDistributionWrapper]]:
    """Constructs the necessary ModelV2 and action dist class for the Policy.

    Args:
        policy: The TFPolicy that will use the models.
        obs_space (gym.spaces.Space): The observation space.
        action_space (gym.spaces.Space): The action space.
        config: The SACConfig object.

    Returns
    -------
        ModelV2: The ModelV2 to be used by the Policy. Note: An additional
            target model will be created in this function and assigned to
            `policy.target_model`.
    """
    model = build_sac_model(policy, obs_space, action_space, config)
    action_dist_class = _get_dist_class(policy, config, action_space)
    return model, action_dist_class


SACModTorchPolicy = build_policy_class(
    name="SACModTorchPolicy",
    framework="torch",
    loss_fn=actor_critic_loss,
    get_default_config=lambda: ray.rllib.algorithms.sac.sac.SACConfig(),
    stats_fn=stats,
    postprocess_fn=postprocess_trajectory,
    extra_grad_process_fn=apply_grad_clipping,
    optimizer_fn=optimizer_fn,
    validate_spaces=validate_spaces,
    before_loss_init=setup_late_mixins,
    make_model_and_action_dist=build_mod_sac_model_and_action_dist,
    extra_learn_fetches_fn=concat_multi_gpu_td_errors,
    mixins=[TargetNetworkMixin, ComputeTDErrorMixin],
    action_distribution_fn=action_distribution_fn,
)


class ModSAC(SAC):
    @classmethod
    @override(DQN)
    def get_default_policy_class(cls, config: AlgorithmConfig) -> type[Policy] | None:
        if config["framework"] == "torch":
            return SACModTorchPolicy
        else:
            return SACTFPolicy


class ModSACConfig(SACConfig):
    def __init__(self, algo_class=None):
        super().__init__(algo_class=ModSAC)


config = SACConfig().framework("torch").environment("Pendulum-v1").debugging(seed=9001)
mod_config = ModSACConfig().framework("torch").environment("Pendulum-v1").debugging(seed=9001)

algorithm = config.build()
mod_algorithm = mod_config.build()


# run for 500 training steps
for _ in range(500):
    result = algorithm.train()
    mod_result = mod_algorithm.train()
    pprint(result)
```

### Tensorboard results (as of current training iteration):

> NOTE:  Green: default SAC implementation | Purple: SAC with exact reparameterization from literature

![image](https://github.com/ray-project/ray/assets/7985947/9bc845d9-caa3-4978-b710-f812002e3bb3)

### Issue Severity

Low: It annoys or frustrates me.

**Comment by simonsays1980** [comment]
@gresavage Thanks for filing this. This is indeed tricky. 

Afaiks the reparameterization trick is only needed for stochastic nodes in the computation graph - so where action sampling happens. At these nodes the `RLlib` implementation uses [`rsample()`](https://pytorch.org/docs/stable/_modules/torch/distributions/normal.html#Normal.rsample) from the `torch.distributions.normal.Normal` and this function already does the reparameterization trick for us. So we get gradients when we backpropagate through this node. 

The difference might come from the difference in the `logp` as this is not the same for `x` and `z=(x-mu)/sigma`, but to be precise I have to go through all steps in the optimization and see where it pops up.
